File: Trabajo pasado/src/icc/ajedrez/Peon.java
package icc.ajedrez;
import java.util.List;
import java.util.ArrayList;

/**Clase para el peón del ajedrez.
*@author Sebastián A.
*/
public class Peon extends Pieza {

	/**Método constructor para el peón.
	*@param columna Columna en la que irá la pieza.
	*@param renglón Renglón en el que irá la pieza.
	*@param color Color de la pieza.
	*/
	public Peon(char columna, int renglón, boolean color) {
		super(columna, renglón, color);
	} //Fin de constructor Peon.

	/**Método para saber todos los posibles movimientos de la pieza.
	*@return Lista con los posibles movimientos de la pieza en la forma (columna,renglón).
	*/
	public List posiblesMovimientos() {

		int posibles = 0;
		int renglónContador = getRenglón();
		char columnaContador = getColumna();
		List<String> listaDePosibles = new ArrayList<String>();

		if (getColor()) { //Si el peón es blanco...

			//NORTE pues es blanco.
			if (renglónContador == 2) { //Si el peón no se ha movido...
				posibles += 2;
				renglónContador += 1;
				listaDePosibles.add("("+columnaContador+","+renglónContador+")");
				renglónContador += 1;
				listaDePosibles.add("("+columnaContador+","+renglónContador+")");
			} /*Fin si el peón no se ha movido. */ else { //Si el peón ya se ha movido...
				if (renglónContador < 8) { //Para evitar contar el renglón 9.
					posibles += 1;
					renglónContador += 1;
					listaDePosibles.add("("+columnaContador+","+renglónContador+")");
				}
			} //Fin si el peón ya se ha movido.

		} /*Fin si el peón es blanco. */ else { //Si el peón es negro...

			//SUR pues es negro.
			if (renglónContador == 7) { //Si el peón no se ha movido...
				posibles += 2;
				renglónContador -= 1;
				listaDePosibles.add("("+columnaContador+","+renglónContador+")");
				renglónContador -= 1;
				listaDePosibles.add("("+columnaContador+","+renglónContador+")");
			} /*Fin si el peón no se ha movido. */ else { //Si el peón ya se ha movido...
				if (renglónContador > 1) { //Para evitar contar el renglón 0.
					posibles += 1;
					renglónContador -= 1;
					listaDePosibles.add("("+columnaContador+","+renglónContador+")");
				}
			} //Fin si el peón ya se ha movido.

		} //Fin si el peón es negro.

		return listaDePosibles;

	} //Fin de posiblesMovimientos().

	/**Método para obtener la pieza representada en forma de String.
	*@return Representación en cadena de la pieza.
	*/
	public String toString() {
		String colorPeón;
		if (getColor()) {colorPeón = "blanco";} else {colorPeón = "negro";}
		return "Este peón "+colorPeón+" se encuentra en la posición: ("+getColumna()+","+getRenglón()+").";
	}

} //Fin de clase Peon.

File: code/piezas/Peon.rb
require_relative '../Pieza'

# Clase que representa a la pieza Peón.
class Peon < Pieza

  # Inicializador de un Peón.
  # columna:: Columna del Peón.
  # renglón:: Renglón del Peón.
  # color:: Color del Peón (booleano).
  def initialize(columna, renglón, color)
    if color && renglón == 1
      raise ArgumentError, "Un peón blanco no puede ubicarse " +
                            "en el primer renglón."
    end

    if !color && renglón == 8
      raise ArgumentError, "Un peón negro no puede ubicarse " +
                            "en el último renglón."
    end

    super(columna, renglón, color);
  end

  def posiblesMovimientos
    # Lista con los posibles movimientos
    # de la pieza.
    listaPM = []

    # Número de veces que se puede desplazar el peón.
    t = (self.posición.renglón == 2 || self.posición.renglón == 7) ? 2 : 1

    begin
      t.times do |z|
        if @color
          listaPM.push( desplazar(0, z+1) )
        else
          listaPM.push( desplazar(0, -(z+1)) )
        end
      end
    rescue
    end

    # Se regresa la lista...
    return listaPM
  end

end